lib/vast/Dialect/HighLevel/Transforms/HLLowerTypes.cpp

```diff
@@ -461,112 +342,85 @@ namespace vast::hl
-        {
-            auto fn = mlir::dyn_cast< FuncOp >(op);
-            if (!fn)
+        // As the reference how to lower functions, the `StandardToLLVM`
+        // conversion is used.
+        //
+        // But basically we need to copy the function with the converted
+        // function type -> copy body -> fix arguments of the entry region.
+        logical_result matchAndRewrite(
+            FuncOp fn, OpAdaptor adaptor, mlir::ConversionPatternRewriter &rewriter
+        ) const override {
+            auto fty = adaptor.getFunctionType();
+            auto &tc = *getTypeConverter();
+
+            signature_conversion_t sigconvert(fty.getNumInputs());
+            if (mlir::failed(tc.convertSignatureArgs(fty.getInputs(), sigconvert))) {
                 return mlir::failure();
+            }
 
-            auto sigconvert = get_fn_signature(*getTypeConverter(), fn, false);
-            if (!sigconvert)
-            {
+            llvm::SmallVector< mlir_type, 1 > results;
+            if (mlir::failed(tc.convertTypes(fty.getResults(), results))) {
                 return mlir::failure();
             }
 
-            // TODO(hllowertypes): What about function attributes?
-            std::vector< mlir::NamedAttribute > attributes;
-            auto arg_attrs = lower_attrs(fn, fn.getArgAttrsAttr());
-            auto res_attrs = lower_attrs(fn, fn.getResAttrsAttr());
-
-            auto maybe_fn_type = getTypeConverter()->convert_type_to_type(fn.getFunctionType());
-            if (!maybe_fn_type)
-                return mlir::failure();
-            auto fn_type = maybe_fn_type->dyn_cast< mlir::FunctionType >();
-            if (!fn_type)
-                return mlir::failure();
+            auto params = sigconvert.getConvertedTypes();
 
-            // Create new function with converted type
-            FuncOp new_fn = rewriter.create< FuncOp >(
-                fn.getLoc(), fn.getName(), fn_type, fn.getLinkage(),
-                attributes, arg_attrs, res_attrs
+            auto new_type = core::FunctionType::get(
+                rewriter.getContext(), params, results, fty.isVarArg()
             );
-            new_fn.setVisibility(mlir::SymbolTable::Visibility::Private);
 
-            // Copy the old region - it will have incorrect arguments (`BlockArgument` on
-            // entry `Block`.
-            rewriter.inlineRegionBefore(fn.getBody(), new_fn.getBody(), new_fn.end());
+            // TODO deal with function attribute types
 
-            // NOTE(lukas): This may break the contract that all modifications happen
-            //              via rewriter.
-            util::convert_region_types(fn, new_fn, *sigconvert);
+            rewriter.updateRootInPlace(fn, [&] {
+                fn.setType(new_type);
+                for (auto [ty, param] : llvm::zip(params, fn.getBody().getArguments())) {
+                    param.setType(ty);
+                }
+            });
```

Comment:
Why not `applySignatureConversion` from `ConversionPatternRewriter`?

Reply:
because it's broken